### PR TITLE
feat: use exactly equal parts in multipart upload

### LINKED ISSUE #4304: Support multipart upload in R2
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

While Minio and S3 are fine with multipart uploads where the parts are of variable size, R2 will error if the parts aren't exactly equal. This has been discussed in https://github.com/apache/arrow/issues/34363 and confirmed with the R2 team that this likely won't be fixed on their end in the near future.

**Describe the solution you'd like**

I think it would be acceptable to change our code base to use even size parts, at least for the time being. I don't think we need to expose a configuration option, unless asked for later.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by wjones127**
# Which issue does this PR close?

Closes #4304.

# Rationale for this change
 
Previously, when a call to `poll_write()` was made, we would write the full buffer provided unless it was too small (when combined with previous values). So if 100MB of data was provided, we would write a part that was 100MB.

This meant that part sizes could be uneven. This is fine with S3 and Minio, but it rejected by R2. In R2, once you upload your first part, you are committing to only uploading parts of exactly equal size (except for the final part).

TBH, I think it's probably an improvement we don't blindly write the full buffer in one go. This should lead to more predictable performance.

# What changes are included in this PR?

The behavior change is described above.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by wjones127** [comment]
✅ Tested locally against Cloudflare R2.

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -1237,8 +1249,8 @@ mod tests {
         assert_eq!(bytes_expected, bytes_written);
 
         // Can overwrite some storage
-        // Sizes carefully chosen to exactly hit min limit of 5 MiB
-        let data = get_vec_of_bytes(242_880, 22);
+        // Sizes chozen to ensure we write three parts
```
Comment (tustvold) [inline object_store/src/lib.rs:1252]:
```suggestion
        // Sizes chosen to ensure we write three parts
```

File: object_store/src/multipart.rs [hunk 2]
```diff
@@ -158,13 +170,18 @@ where
...
+            offset += self.as_mut().add_to_buffer(buf, offset);
+
+            // If we don't have a full buffer or we have too many tasks, break
+            if self.current_buffer.len() < self.part_size
+                || self.tasks.len() >= self.max_concurrency
+            {
+                break;
+            }
 
             let out_buffer = std::mem::take(&mut self.current_buffer);
```
Comment (tustvold) [inline object_store/src/multipart.rs:186]:
```suggestion
            let out_buffer = std::mem::replace(&mut self.current_buffer, Vec::with_capacity(self.part_size));
```
Perhaps

**Review by tustvold — APPROVED** [review]
Looks good to me

**Pull request merged**